lib/telemetry_metrics.ex

```diff
@@ -1,5 +1,390 @@
+  every flush and can be decremented, whereas counters in Prometheus are monotonically increasing.
+  `Telemetry.Metrics` doesn't focus on those details - instead, it describes what the end user,
+  operator, expects to see when using the metric of particular type. This implies that in most
+  cases aggregated metrics won't be visible inside the BEAM, but in exchange aggregations can be
+  implemented in a way that makes most sense for particular system. Finally, one could also
+  implement an in-VM "reporter" which would aggregate the metrics and expose them inside the BEAM.
+  When there is a need to swap the reporters, and if both reporters are following the metric types
+  specification, then the end result of aggregation is the same, regardless of the backend system
+  in use.
+
+  ### Requirements for reporters
+
+  Reporters should accept metric specifications and subscribe to relevant events. When those events
+  are emitted, metric should be updated (either in-memory or by contacting external system) in such
+  a way that the user is able to view metric values as described in the "Metric types" section.
+
+  If the reporter does not support the metric given to it, it should log a warning.
+
+  Reporters should also document how `Telemetry.Metrics` metric types, names tags are translated to
+  metric types and identifiers in the system they publish metrics to.
   """
+
+  @type event_name :: String.t() | Telemetry.event_name()
+  @type metric_name :: String.t() | normalized_metric_name()
+  @type normalized_metric_name :: [atom(), ...]
+  @type metric_type :: :counter | :sum | :last_value | :distribution
+  @type metadata ::
+          :all | [key :: term()] | (Telemetry.event_metadata() -> Telemetry.event_metadata())
+  @type tag :: term()
+  @type tags :: [tag()]
+  @type description :: nil | String.t()
+  @type unit :: atom()
+  @type counter_options :: [metric_option()]
+  @type sum_options :: [metric_option()]
+  @type last_value_options :: [metric_option()]
+  @type metric_option ::
+          {:name, metric_name()}
+          | {:metadata, metadata()}
+          | {:tags, tags()}
+          | {:description, description()}
+          | {:unit, unit()}
+
+  require Logger
+
+  alias Telemetry.Metrics.{Counter, Sum, LastValue}
+
+  # API
+
+  @doc """
+  Returns a specification of counter metric.
+
+  See "Metric specifications" section in the top-level documentation of this module for more
+  information.
+
+  ## Example
+
+      counter(
+        "http.request",
+        name: "http.request.count", metadata: [:controller, :action] tags: [:controller, :action]
+      )
+  """
+  @spec counter(Telemetry.event_name(), counter_options()) :: Counter.t()
+  def counter(event_name, options) do
+    validate_event_name!(event_name)
+    validate_metric_options!(options)
+    options = Keyword.merge(default_metric_options(event_name), options)
+
+    event_name = normalize_event_or_metric_name(event_name)
```

Comment:
Instead of having `normalize_event_or_metric_name`, you could have `validate_event_or_metric_name!()` and make sure it returns the normalized format, so you don't have to traverse it twice. :)